panther_hardware_interfaces/src/panther_imu.cpp

```diff
@@ -0,0 +1,542 @@
+      "work");
+  }
+}
+
+void PantherImuSensor::ReadCompassParams()
+{
+  params_.cc_mag_field = hardware_interface::stod(info_.hardware_parameters.at("cc_mag_field"));
+  params_.cc_offset0 = hardware_interface::stod(info_.hardware_parameters.at("cc_offset0"));
+  params_.cc_offset1 = hardware_interface::stod(info_.hardware_parameters.at("cc_offset1"));
+  params_.cc_offset2 = hardware_interface::stod(info_.hardware_parameters.at("cc_offset2"));
+  params_.cc_gain0 = hardware_interface::stod(info_.hardware_parameters.at("cc_gain0"));
+  params_.cc_gain1 = hardware_interface::stod(info_.hardware_parameters.at("cc_gain1"));
+  params_.cc_gain2 = hardware_interface::stod(info_.hardware_parameters.at("cc_gain2"));
+  params_.cc_t0 = hardware_interface::stod(info_.hardware_parameters.at("cc_t0"));
+  params_.cc_t1 = hardware_interface::stod(info_.hardware_parameters.at("cc_t1"));
+  params_.cc_t2 = hardware_interface::stod(info_.hardware_parameters.at("cc_t2"));
+  params_.cc_t3 = hardware_interface::stod(info_.hardware_parameters.at("cc_t3"));
+  params_.cc_t4 = hardware_interface::stod(info_.hardware_parameters.at("cc_t4"));
+  params_.cc_t5 = hardware_interface::stod(info_.hardware_parameters.at("cc_t5"));
+}
+
+void PantherImuSensor::ReadMadgwickFilterParams()
+{
+  params_.gain = hardware_interface::stod(info_.hardware_parameters.at("gain"));
+  params_.zeta = hardware_interface::stod(info_.hardware_parameters.at("zeta"));
+  params_.mag_bias_x = hardware_interface::stod(info_.hardware_parameters.at("mag_bias_x"));
+  params_.mag_bias_y = hardware_interface::stod(info_.hardware_parameters.at("mag_bias_y"));
+  params_.mag_bias_z = hardware_interface::stod(info_.hardware_parameters.at("mag_bias_z"));
+  params_.use_mag = hardware_interface::parse_bool(info_.hardware_parameters.at("use_mag"));
+  params_.stateless = hardware_interface::parse_bool(info_.hardware_parameters.at("stateless"));
+  params_.remove_gravity_vector =
+    hardware_interface::parse_bool(info_.hardware_parameters.at("remove_gravity_vector"));
+
+  CheckMadgwickFilterWorldFrameParam();
+}
+
+void PantherImuSensor::CheckMadgwickFilterWorldFrameParam()
+{
+  const auto world_frame = info_.hardware_parameters.at("world_frame");
+
+  if (world_frame == "ned") {
+    world_frame_ = WorldFrame::NED;
+  } else if (world_frame == "nwu") {
+    world_frame_ = WorldFrame::NWU;
+  } else if (world_frame == "enu") {
+    world_frame_ = WorldFrame::ENU;
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "The parameter world_frame was set to invalid value. "
+      "Valid values are 'enu', 'ned' and 'nwu'. Setting to 'enu'.");
+  }
+}
+
+void PantherImuSensor::SetInitialValues()
+{
+  imu_sensor_state_.resize(
+    info_.sensors[0].state_interfaces.size(), std::numeric_limits<double>::quiet_NaN());
+}
+
+void PantherImuSensor::Calibrate()
+{
+  spatial_->zero();
+
+  RCLCPP_WARN(logger_, "IMU is calibrating. Do not move the robot for 2 seconds!");
+  while (!imu_calibrated_) {
```

Comment:
Should we use IsIMUCalibrated here? Also, I would like to inquire again if it's feasible to wait until the IMU is calibrated using synchronization mechanisms such as std::condition_variable or std::promise. I believe it is possible :p